File: src/v3/engine.ts
import { WeightedLruCache } from './cache.js';
import {
  compileDiceProgram,
  compilePreparedDicePlan,
  hasPlanProgram,
  prepareDicePlanInput,
  validateKnownPlan,
  type CompiledDiceProgram,
} from './compiler.js';
import { DiceRollError, isDiceRollError } from './errors.js';
import { executeRollPlan, executeRollPlanSummary } from './executor.js';
import {
  freezeDiceRollResult,
  freezeDiceRollSummary,
  shouldFreezeResults,
} from './freeze.js';
import { normalizeRpgDiceNotation, parseNormalizedDiceInput } from './normalization.js';
import {
  createDiceLimits,
  resolveDiceLimits,
  type DiceLimits,
} from './runtime/limits.js';
import {
  validateReplayDescriptor,
  type RandomAlgorithm,
  type ReplayDescriptor,
  type SeedInput,
} from './runtime/replay.js';
import type {
  CompileOptions,
  DiceCacheOptions,
  DiceCacheStats,
  DiceEngine,
  DiceEngineOptions,
  DiceNotationInspection,
  DiceRollResult,
  DiceRollSummary,
  FreezeResultsMode,
  RollOptions,
  RollPlan,
} from './types.js';

const DEFAULT_INPUT_CACHE_ENTRIES = 500;
const DEFAULT_PROGRAM_CACHE_ENTRIES = 200;
const DEFAULT_PROGRAM_CACHE_NODES = 100_000;
const UNBOUNDED_ENTRY_WEIGHT = Number.MAX_SAFE_INTEGER;

interface ResolvedCacheOptions {
  readonly maxInputEntries: number;
  readonly maxProgramEntries: number;
  readonly maxProgramNodes: number;
}

interface ExternalPlanEnvelope {
  readonly input: string;
  readonly planFingerprint: string;
}

interface RuntimeRollOptions {
  readonly limits: unknown;
  readonly randomAlgorithm: RandomAlgorithm | undefined;
  readonly replay: unknown;
  readonly seed: SeedInput | undefined;
}

interface RuntimeEngineOptions {
  readonly cache: unknown;
  readonly freezeResults: FreezeResultsMode | undefined;
  readonly limits: unknown;
  readonly randomAlgorithm: RandomAlgorithm | undefined;
}

function validateCacheLimit(name: keyof DiceCacheOptions, value: unknown, fallback: number): number {
  const resolved = value ?? fallback;
  if (typeof resolved !== 'number' || !Number.isSafeInteger(resolved) || resolved < 0) {
    throw new RangeError(`${name} must be a non-negative safe integer`);
  }
  return resolved;
}

function resolveCacheOptions(options: unknown): ResolvedCacheOptions {
  if (options === false) {
    return { maxInputEntries: 0, maxProgramEntries: 0, maxProgramNodes: 0 };
  }
  let cache: object;
  try {
    if (options !== undefined && (!isObject(options) || Array.isArray(options))) {
      throw new RangeError('cache must be false or a cache options object');
    }
    cache = options === undefined ? {} : options;
  } catch (error: unknown) {
    if (error instanceof RangeError) {
      throw error;
    }
    throw new RangeError('cache options could not be read');
  }
  let maxInputEntries: unknown;
  let maxProgramEntries: unknown;
  let maxProgramNodes: unknown;
  try {
    maxInputEntries = readProperty(cache, 'maxInputEntries');
    maxProgramEntries = readProperty(cache, 'maxProgramEntries');
    maxProgramNodes = readProperty(cache, 'maxProgramNodes');
  } catch {
    throw new RangeError('cache options could not be read');
  }
  return {
    maxInputEntries: validateCacheLimit(
      'maxInputEntries', maxInputEntries, DEFAULT_INPUT_CACHE_ENTRIES,
    ),
    maxProgramEntries: validateCacheLimit(
      'maxProgramEntries', maxProgramEntries, DEFAULT_PROGRAM_CACHE_ENTRIES,
    ),
    maxProgramNodes: validateCacheLimit(
      'maxProgramNodes', maxProgramNodes, DEFAULT_PROGRAM_CACHE_NODES,
    ),
  };
}

function limitsCacheKey(limits: DiceLimits): string {
  return [
    limits.maxInputLength,
    limits.maxAstDepth,
    limits.maxAstNodes,
    limits.maxRolls,
    limits.maxInitialDice,
    limits.maxSides,
    limits.maxModifierSteps,
  ].join(':');
}

function inputCacheKey(input: string, limits: DiceLimits): string {
  return `${limitsCacheKey(limits)}\u0000${input}`;
}

function isObject(value: unknown): value is object {
  return typeof value === 'object' && value !== null;
}

function hasProperty(value: object, key: string): value is object & Readonly<Record<string, unknown>> {
  return key in value;
}

function readProperty(value: object, key: string): unknown {
  return hasProperty(value, key) ? value[key] : undefined;
}

function readOptionsProperty(options: unknown, key: string, code: 'INVALID_LIMIT' | 'INVALID_REPLAY'): unknown {
  try {
    if (!isObject(options) || Array.isArray(options)) {
      throw new DiceRollError('Dice options must be a non-null object', { code });
    }
    return readProperty(options, key);
  } catch (error: unknown) {
    if (isDiceRollError(error)) {
      throw error;
    }
    throw new DiceRollError(`Unable to read dice option ${key}`, {
      code,
      details: { option: key },
    });
  }
}

function readCompileLimits(options: unknown): unknown {
  return readOptionsProperty(options, 'limits', 'INVALID_LIMIT');
}

function readRuntimeRollOptions(options: unknown): RuntimeRollOptions {
  const limits = readOptionsProperty(options, 'limits', 'INVALID_LIMIT');
  const seed = readOptionsProperty(options, 'seed', 'INVALID_REPLAY');
  const replay = readOptionsProperty(options, 'replay', 'INVALID_REPLAY');
  const randomAlgorithm = readOptionsProperty(options, 'randomAlgorithm', 'INVALID_REPLAY');
  if (seed !== undefined && typeof seed !== 'string' && typeof seed !== 'number') {
    throw new DiceRollError('Seeds must be finite numbers or strings', { code: 'INVALID_SEED' });
  }
  if (randomAlgorithm !== undefined
    && randomAlgorithm !== 'mt19937'
    && randomAlgorithm !== 'xoshiro128ss') {
    throw new DiceRollError('The requested random algorithm is not supported', {
      code: 'INVALID_REPLAY',
    });
  }
  if (replay !== undefined && (seed !== undefined || randomAlgorithm !== undefined)) {
    throw new DiceRollError('Replay cannot be combined with seed or randomAlgorithm', {
      code: 'INVALID_REPLAY',
    });
  }
  return { limits, randomAlgorithm, replay, seed };
}

function readRuntimeEngineOptions(options: unknown): RuntimeEngineOptions {
  const limits = readOptionsProperty(options, 'limits', 'INVALID_LIMIT');
  const cache = readOptionsProperty(options, 'cache', 'INVALID_LIMIT');
  const freezeResults = readOptionsProperty(options, 'freezeResults', 'INVALID_LIMIT');
  const randomAlgorithm = readOptionsProperty(options, 'randomAlgorithm', 'INVALID_REPLAY');
  if (freezeResults !== undefined
    && freezeResults !== 'development'
    && freezeResults !== 'always'
    && freezeResults !== 'never') {
    throw new DiceRollError('freezeResults uses an unsupported mode', { code: 'INVALID_LIMIT' });
  }
  if (randomAlgorithm !== undefined
    && randomAlgorithm !== 'mt19937'
    && randomAlgorithm !== 'xoshiro128ss') {
    throw new DiceRollError('The requested random algorithm is not supported', {
      code: 'INVALID_REPLAY',
    });
  }
  return { cache, freezeResults, limits, randomAlgorithm };
}

function resolveCallLimits(engineLimits: DiceLimits, overrides: unknown): DiceLimits {
  return overrides === undefined
    ? engineLimits
    : resolveDiceLimits(engineLimits, overrides);
}

function replayForPlan(replay: unknown, plan: RollPlan): ReplayDescriptor {
  return validateReplayDescriptor(replay, plan.planFingerprint);
}

function readExternalPlan(value: unknown): ExternalPlanEnvelope {
  try {
    if (!isObject(value)
      || readProperty(value, 'type') !== 'roll-plan'
      || readProperty(value, 'schemaVersion') !== 3
      || readProperty(value, 'compilerVersion') !== 1) {
      throw new DiceRollError('Roll plan is invalid or uses an unsupported schema', {
        code: 'UNSUPPORTED_NOTATION',
      });
    }
    const input = readProperty(value, 'input');
    const planFingerprint = readProperty(value, 'planFingerprint');
    if (typeof input !== 'string' || typeof planFingerprint !== 'string'
      || planFingerprint.length !== 32
      || !/^[0-9a-f]{32}$/u.test(planFingerprint)) {
      throw new DiceRollError('Roll plan envelope is malformed', {
        code: 'UNSUPPORTED_NOTATION',
      });
    }
    return { input, planFingerprint };
  } catch (error: unknown) {
    if (isDiceRollError(error)) {
      throw error;
    }
    throw new DiceRollError('Roll plan could not be read safely', {
      code: 'UNSUPPORTED_NOTATION',
      details: { cause: error instanceof Error ? error.message : 'External plan access failed' },
    });
  }
}

function validInspection(plan: RollPlan): DiceNotationInspection {
  return {
    type: 'dice-notation-inspection',
    input: plan.input,
    notation: plan.notation,
    normalizedNotation: plan.normalizedNotation,
    comment: plan.comment,
    isValid: true,
    plan,
    groups: plan.groups,
    cost: plan.cost,
    error: null,
  };
}

function invalidInspection(input: string, error: unknown): DiceNotationInspection {
  const safeInput = typeof input === 'string' ? input : '';
  const normalizedError = isDiceRollError(error)
    ? error
    : new DiceRollError('Invalid dice notation', {
        code: 'INVALID_NOTATION',
        input: safeInput,
        details: { cause: error instanceof Error ? error.message : 'Unknown compiler error' },
      });
  const normalized = normalizedError.code === 'INPUT_TOO_LONG'
    ? {
        input: safeInput,
        notation: '',
        normalizedNotation: '',
        comment: '',
        rollCount: 1,
        isMultiRoll: false,
      }
    : parseNormalizedDiceInput(safeInput);
  return {
    type: 'dice-notation-inspection',
    input: safeInput,
    notation: normalized.notation,
    normalizedNotation: normalized.normalizedNotation,
    comment: normalized.comment,
    isValid: false,
    plan: null,
    groups: [],
    cost: null,
    error: normalizedError,
  };
}

class DefaultDiceEngine implements DiceEngine {
  readonly limits: DiceLimits;

  private readonly freezeResults: FreezeResultsMode;

  private readonly randomAlgorithm: RandomAlgorithm;

  private readonly inputCache: WeightedLruCache<string, RollPlan>;

  private readonly programCache: WeightedLruCache<string, CompiledDiceProgram>;

  private readonly ownedPlans = new WeakSet<RollPlan>();

  constructor(options: DiceEngineOptions) {
    const runtimeOptions = readRuntimeEngineOptions(options);
    this.limits = createDiceLimits(runtimeOptions.limits ?? {});
    this.freezeResults = runtimeOptions.freezeResults ?? 'never';
    this.randomAlgorithm = runtimeOptions.randomAlgorithm ?? 'mt19937';
    const cache = resolveCacheOptions(runtimeOptions.cache);
    this.inputCache = new WeightedLruCache(
      cache.maxInputEntries,
      UNBOUNDED_ENTRY_WEIGHT,
    );
    this.programCache = new WeightedLruCache(
      cache.maxProgramEntries,
      cache.maxProgramNodes,
    );
  }

  clearCache(): void {
    this.inputCache.clear();
    this.programCache.clear();
  }

  getCacheStats(): DiceCacheStats {
    const input = this.inputCache.stats();
    const program = this.programCache.stats();
    return {
      inputEntries: input.entries,
      programEntries: program.entries,
      programNodes: program.weight,
      hits: input.hits + program.hits,
      misses: input.misses + program.misses,
      evictions: input.evictions + program.evictions,
    };
  }

  compile(input: string, options: CompileOptions = {}): RollPlan {
    if (typeof input !== 'string') {
      throw new DiceRollError('Dice input must be a string', {
        code: 'INVALID_NOTATION',
        details: { receivedType: typeof input },
      });
    }
    const limits = resolveCallLimits(this.limits, readCompileLimits(options));
    if (input.length > limits.maxInputLength) {
      throw new DiceRollError('Dice input exceeds the configured length limit', {
        code: 'INPUT_TOO_LONG',
        input,
        details: { actual: input.length, limit: limits.maxInputLength },
      });
    }
    const key = inputCacheKey(input, limits);
    const cached = this.inputCache.get(key);
    if (cached !== undefined) {
      return cached;
    }

    const prepared = prepareDicePlanInput(input, limits);
    const programKey = prepared.normalized.notation;
    let program = this.programCache.get(programKey);
    if (program === undefined) {
      program = compileDiceProgram(programKey, input, limits);
      this.programCache.set(programKey, program, program.nodeCount);
    }
    const plan = compilePreparedDicePlan(prepared, limits, program);
    this.ownedPlans.add(plan);
    this.inputCache.set(key, plan);
    return plan;
  }

  inspect(input: string, options: CompileOptions = {}): DiceNotationInspection {
    try {
      return validInspection(this.compile(input, options));
    } catch (error: unknown) {
      return invalidInspection(input, error);
    }
  }

  normalize(input: string): string {
    return normalizeRpgDiceNotation(input);
  }

  roll(input: string | RollPlan, options: RollOptions = {}): DiceRollResult {
    const runtimeOptions = readRuntimeRollOptions(options);
    const limits = resolveCallLimits(this.limits, runtimeOptions.limits);
    const plan = this.resolvePlan(input, limits);
    const result = runtimeOptions.replay !== undefined
      ? executeRollPlan(plan, { limits, replay: replayForPlan(runtimeOptions.replay, plan) })
      : runtimeOptions.seed === undefined
        ? executeRollPlan(plan, {
            limits,
            randomAlgorithm: runtimeOptions.randomAlgorithm ?? this.randomAlgorithm,
          })
        : executeRollPlan(plan, {
            limits,
            seed: runtimeOptions.seed,
            randomAlgorithm: runtimeOptions.randomAlgorithm ?? this.randomAlgorithm,
          });
    return shouldFreezeResults(this.freezeResults) ? freezeDiceRollResult(result) : result;
  }

  rollSummary(input: string | RollPlan, options: RollOptions = {}): DiceRollSummary {
    const runtimeOptions = readRuntimeRollOptions(options);
    const limits = resolveCallLimits(this.limits, runtimeOptions.limits);
    const plan = this.resolvePlan(input, limits);
    const result = runtimeOptions.replay !== undefined
      ? executeRollPlanSummary(plan, {
          limits,
          replay: replayForPlan(runtimeOptions.replay, plan),
        })
      : runtimeOptions.seed === undefined
        ? executeRollPlanSummary(plan, {
            limits,
            randomAlgorithm: runtimeOptions.randomAlgorithm ?? this.randomAlgorithm,
          })
        : executeRollPlanSummary(plan, {
            limits,
            seed: runtimeOptions.seed,
            randomAlgorithm: runtimeOptions.randomAlgorithm ?? this.randomAlgorithm,
          });
    return shouldFreezeResults(this.freezeResults) ? freezeDiceRollSummary(result) : result;
  }

  verify(input: string, options: CompileOptions = {}): boolean {
    return this.inspect(input, options).isValid;
  }

  private resolvePlan(input: string | RollPlan, limits: DiceLimits): RollPlan {
    if (typeof input === 'string') {
      return this.compile(input, { limits });
    }
    if (hasPlanProgram(input)) {
      if (!this.ownedPlans.has(input) || limits !== this.limits) {
        validateKnownPlan(input, limits);
      }
      return input;
    }
    const envelope = readExternalPlan(input);
    const plan = this.compile(envelope.input, { limits });
    if (envelope.planFingerprint !== plan.planFingerprint) {
      throw new DiceRollError('Roll plan does not match its source input', {
        code: 'UNSUPPORTED_NOTATION',
        input: envelope.input,
        details: {
          expectedFingerprint: plan.planFingerprint,
          receivedFingerprint: envelope.planFingerprint,
        },
      });
    }
    return plan;
  }
}

export function createDiceEngine(options: DiceEngineOptions = {}): DiceEngine {
  return new DefaultDiceEngine(options);
}

const defaultDiceEngine = createDiceEngine();

export function compileRpgDice(input: string, options: CompileOptions = {}): RollPlan {
  return defaultDiceEngine.compile(input, options);
}

export function inspectRpgDiceNotation(
  input: string,
  options: CompileOptions = {},
): DiceNotationInspection {
  return defaultDiceEngine.inspect(input, options);
}

export function rollRpgDice(
  input: string | RollPlan,
  options: RollOptions = {},
): DiceRollResult {
  return defaultDiceEngine.roll(input, options);
}

export function rollRpgDiceSummary(
  input: string | RollPlan,
  options: RollOptions = {},
): DiceRollSummary {
  return defaultDiceEngine.rollSummary(input, options);
}

export function verifyRpgDiceNotation(input: string, options: CompileOptions = {}): boolean {
  return defaultDiceEngine.verify(input, options);
}

File: go/engine.go
package dicecore

import (
	"encoding/json"
	"errors"
	"fmt"
	"sync"
)

// Engine owns bounded compilation caches. Rolls have independent execution state
// and may run concurrently. Public plans and results are caller-owned values.
type Engine struct {
	limits       DiceLimits
	algorithm    RandomAlgorithm
	freezeMode   FreezeResultsMode
	mu           sync.Mutex
	inputCache   *WeightedLruCache[engineInputKey, *RollPlan]
	programCache *WeightedLruCache[string, *CompiledDiceProgram]
}

func resolveEngineCache(options any) ([3]int64, error) {
	caps := [3]int64{500, 200, 100000}
	if disabled, ok := options.(bool); ok && !disabled {
		return [3]int64{}, nil
	}
	var overrides DiceCacheOptions
	switch value := options.(type) {
	case nil:
	case DiceCacheOptions:
		overrides = value
	default:
		return caps, fmt.Errorf("cache must be false or a cache options object")
	}
	for index, key := range []string{"maxInputEntries", "maxProgramEntries", "maxProgramNodes"} {
		if value, ok := overrides[key]; ok {
			if value < 0 || value > maxSafeInteger {
				return caps, fmt.Errorf("%s must be a non-negative safe integer", key)
			}
			caps[index] = value
		}
	}
	return caps, nil
}

func CreateDiceEngine(options ...DiceEngineOptions) (*Engine, error) {
	if len(options) > 1 {
		return nil, newDiceError("INVALID_LIMIT", "Only one engine options value may be provided", "", nil)
	}
	var option DiceEngineOptions
	if len(options) == 1 {
		option = options[0]
	}
	if option.FreezeResults != "" && option.FreezeResults != "always" && option.FreezeResults != "never" && option.FreezeResults != "development" {
		return nil, newDiceError("INVALID_LIMIT", "freezeResults uses an unsupported mode", "", nil)
	}
	if option.RandomAlgorithm != "" && !supportedRandomAlgorithm(option.RandomAlgorithm) {
		return nil, newDiceError("INVALID_REPLAY", "The requested random algorithm is not supported", "", nil)
	}
	limits, err := CreateDiceLimits(option.Limits)
	if err != nil {
		return nil, err
	}
	caps, err := resolveEngineCache(option.Cache)
	if err != nil {
		return nil, err
	}
	algorithm := option.RandomAlgorithm
	if algorithm == "" {
		algorithm = MT19937
	}
	freeze := option.FreezeResults
	if freeze == "" {
		freeze = "never"
	}
	return &Engine{limits: limits, algorithm: algorithm, freezeMode: freeze,
		inputCache:   NewWeightedLruCache[engineInputKey, *RollPlan](caps[0], maxSafeInteger),
		programCache: NewWeightedLruCache[string, *CompiledDiceProgram](caps[1], caps[2])}, nil
}

var sharedDiceEngine = sync.OnceValue(func() *Engine {
	engine, _ := CreateDiceEngine()
	return engine
})

func DefaultDiceEngine() *Engine          { return sharedDiceEngine() }
func (engine *Engine) Limits() DiceLimits { return engine.limits }

func (engine *Engine) ClearCache() {
	engine.mu.Lock()
	defer engine.mu.Unlock()
	engine.inputCache.Clear()
	engine.programCache.Clear()
}

func (engine *Engine) GetCacheStats() DiceCacheStats {
	engine.mu.Lock()
	defer engine.mu.Unlock()
	input, program := engine.inputCache.Stats(), engine.programCache.Stats()
	return DiceCacheStats{InputEntries: input.Entries, ProgramEntries: program.Entries, ProgramNodes: program.Weight,
		Hits: input.Hits + program.Hits, Misses: input.Misses + program.Misses, Evictions: input.Evictions + program.Evictions}
}

// A native comparable key avoids formatting seven integers and allocating a
// concatenated string for every cache hit. The dimensions match the compiler.
type engineInputKey struct {
	input  string
	limits [7]int64
}

func engineLimitsKey(input string, limits DiceLimits) engineInputKey {
	return engineInputKey{input, [7]int64{limits.MaxInputLength, limits.MaxAstDepth, limits.MaxAstNodes,
		limits.MaxRolls, limits.MaxInitialDice, limits.MaxSides, limits.MaxModifierSteps}}
}

// cloneEnginePlan keeps the cache's canonical plan private while preserving the
// compiler binding. Caller mutation cannot affect later compilations or rolls.
func cloneEnginePlan(plan *RollPlan) *RollPlan {
	clone := *plan
	clone.Groups = make([]RollPlanGroup, len(plan.Groups))
	for index, group := range plan.Groups {
		clone.Groups[index] = group
		clone.Groups[index].ChildIDs = append([]string{}, group.ChildIDs...)
	}
	clone.identity = &clone
	clone.canonical = plan
	return &clone
}

func (engine *Engine) Compile(input string, options ...CompileOptions) (*RollPlan, error) {
	if len(options) > 1 {
		return nil, newDiceError("INVALID_LIMIT", "Only one compile options value may be provided", "", nil)
	}
	var overrides DiceLimitOverrides
	if len(options) == 1 {
		overrides = options[0].Limits
	}
	limits, err := engine.resolveLimits(overrides)
	if err != nil {
		return nil, err
	}
	plan, err := engine.compileResolved(input, limits)
	if err != nil {
		return nil, err
	}
	return cloneEnginePlan(plan), nil
}

func (engine *Engine) resolveLimits(overrides DiceLimitOverrides) (DiceLimits, error) {
	// Constructor-owned limits are already validated and immutable. Keep the
	// full validation path for per-call overrides and an uninitialized Engine.
	if len(overrides) == 0 && engine.inputCache != nil {
		return engine.limits, nil
	}
	return ResolveDiceLimits(engine.limits, overrides)
}

func (engine *Engine) compileResolved(input string, limits DiceLimits) (*RollPlan, error) {
	length := int64(len(syntaxUnits(input)))
	if length > limits.MaxInputLength {
		return nil, newDiceError("INPUT_TOO_LONG", "Dice input exceeds the configured length limit", input,
			map[string]any{"actual": length, "limit": limits.MaxInputLength})
	}
	engine.mu.Lock()
	defer engine.mu.Unlock()
	key := engineLimitsKey(input, limits)
	if cached, ok := engine.inputCache.Get(key); ok {
		return cached, nil
	}
	prepared, err := PrepareDicePlanInput(input, limits)
	if err != nil {
		return nil, err
	}
	programKey := prepared.Normalized.Notation
	program, ok := engine.programCache.Get(programKey)
	if !ok {
		program, err = CompileDiceProgram(programKey, input, limits)
		if err != nil {
			return nil, err
		}
		engine.programCache.Set(programKey, program, program.NodeCount)
	}
	plan, err := BindDicePlan(prepared, program, limits)
	if err != nil {
		return nil, err
	}
	engine.inputCache.Set(key, plan)
	return plan, nil
}

func (engine *Engine) Inspect(input string, options ...CompileOptions) *DiceNotationInspection {
	plan, err := engine.Compile(input, options...)
	if err == nil {
		cost := plan.Cost
		return &DiceNotationInspection{Type: "dice-notation-inspection", Input: plan.Input, Notation: plan.Notation,
			NormalizedNotation: plan.NormalizedNotation, Comment: plan.Comment, IsValid: true, Plan: plan,
			Groups: plan.Groups, Cost: &cost}
	}
	return invalidEngineInspection(input, err)
}

func invalidEngineInspection(input string, err error) *DiceNotationInspection {
	var diceErr *DiceRollError
	if !errors.As(err, &diceErr) {
		diceErr = newDiceError("INVALID_NOTATION", "Invalid dice notation", input, map[string]any{"cause": err.Error()})
	}
	normalized := NormalizedDiceInput{Input: input, RollCount: 1}
	if diceErr.Code != "INPUT_TOO_LONG" {
		normalized = ParseNormalizedInput(input)
	}
	return &DiceNotationInspection{Type: "dice-notation-inspection", Input: input, Notation: normalized.Notation,
		NormalizedNotation: normalized.NormalizedNotation, Comment: normalized.Comment, Groups: []RollPlanGroup{}, Error: diceErr}
}

func (engine *Engine) Normalize(input string) string { return NormalizeRPGDiceNotation(input) }
func (engine *Engine) Verify(input string, options ...CompileOptions) bool {
	return engine.Inspect(input, options...).IsValid
}

func readEngineRollOptions(options []RollOptions) (RollOptions, error) {
	if len(options) > 1 {
		return RollOptions{}, newDiceError("INVALID_REPLAY", "Only one roll options value may be provided", "", nil)
	}
	var option RollOptions
	if len(options) == 1 {
		option = options[0]
	}
	if option.Seed != nil {
		switch option.Seed.(type) {
		case string, UTF16Seed:
		default:
			if _, ok := seedNumber(option.Seed); !ok {
				return option, newDiceError("INVALID_SEED", "Seeds must be finite numbers or strings", "", nil)
			}
		}
	}
	if option.RandomAlgorithm != "" && !supportedRandomAlgorithm(option.RandomAlgorithm) {
		return option, newDiceError("INVALID_REPLAY", "The requested random algorithm is not supported", "", nil)
	}
	if option.Replay != nil && (option.Seed != nil || option.RandomAlgorithm != "") {
		return option, newDiceError("INVALID_REPLAY", "Replay cannot be combined with seed or randomAlgorithm", "", nil)
	}
	return option, nil
}

func (engine *Engine) prepareRoll(input any, options []RollOptions) (*RollPlan, ExecuteRollPlanOptions, error) {
	option, err := readEngineRollOptions(options)
	if err != nil {
		return nil, ExecuteRollPlanOptions{}, err
	}
	limits, err := engine.resolveLimits(option.Limits)
	if err != nil {
		return nil, ExecuteRollPlanOptions{}, err
	}
	plan, err := engine.resolvePlan(input, limits)
	if err != nil {
		return nil, ExecuteRollPlanOptions{}, err
	}
	algorithm := option.RandomAlgorithm
	if option.Replay == nil && algorithm == "" {
		algorithm = engine.algorithm
	}
	return plan, ExecuteRollPlanOptions{Limits: limits, Seed: option.Seed, Replay: option.Replay, RandomAlgorithm: algorithm}, nil
}

func (engine *Engine) Roll(input any, options ...RollOptions) (*DiceRollResult, error) {
	plan, execution, err := engine.prepareRoll(input, options)
	if err != nil {
		return nil, err
	}
	return ExecuteRollPlan(plan, execution)
}

func (engine *Engine) RollDetails(input any, options ...RollOptions) (*DiceRollDetails, error) {
	plan, execution, err := engine.prepareRoll(input, options)
	if err != nil {
		return nil, err
	}
	return ExecuteRollPlanDetails(plan, execution)
}

func (engine *Engine) RollSummary(input any, options ...RollOptions) (*DiceRollSummary, error) {
	plan, execution, err := engine.prepareRoll(input, options)
	if err != nil {
		return nil, err
	}
	return ExecuteRollPlanSummary(plan, execution)
}

func readExternalRollPlan(input any) (string, string, error) {
	var envelope struct {
		Type            string          `json:"type"`
		SchemaVersion   int             `json:"schemaVersion"`
		CompilerVersion int             `json:"compilerVersion"`
		Input           json.RawMessage `json:"input"`
		PlanFingerprint json.RawMessage `json:"planFingerprint"`
	}
	// Inspect only the trusted envelope; unrelated data may contain values that
	// cannot be serialized and is never executable program state.
	if object, ok := input.(map[string]any); ok {
		input = map[string]any{"type": object["type"], "schemaVersion": object["schemaVersion"],
			"compilerVersion": object["compilerVersion"], "input": object["input"], "planFingerprint": object["planFingerprint"]}
	}
	data, err := json.Marshal(input)
	if err != nil {
		return "", "", newDiceError("UNSUPPORTED_NOTATION", "Roll plan could not be read safely", "", map[string]any{"cause": err.Error()})
	}
	if json.Unmarshal(data, &envelope) != nil || envelope.Type != "roll-plan" || envelope.SchemaVersion != 3 || envelope.CompilerVersion != 1 {
		return "", "", newDiceError("UNSUPPORTED_NOTATION", "Roll plan is invalid or uses an unsupported schema", "", nil)
	}
	var notation, fingerprint *string
	if json.Unmarshal(envelope.Input, &notation) != nil || json.Unmarshal(envelope.PlanFingerprint, &fingerprint) != nil ||
		notation == nil || fingerprint == nil || !validReplayHex(*fingerprint) {
		return "", "", newDiceError("UNSUPPORTED_NOTATION", "Roll plan envelope is malformed", "", nil)
	}
	return *notation, *fingerprint, nil
}

func (engine *Engine) resolvePlan(input any, limits DiceLimits) (*RollPlan, error) {
	if notation, ok := input.(string); ok {
		return engine.compileResolved(notation, limits)
	}
	if plan, ok := input.(*RollPlan); ok && HasPlanProgram(plan) {
		if plan.canonical != nil {
			plan = plan.canonical
		}
		if err := ValidatePlanLimits(plan, limits); err != nil {
			return nil, err
		}
		return plan, nil
	}
	notation, fingerprint, err := readExternalRollPlan(input)
	if err != nil {
		return nil, err
	}
	plan, err := engine.compileResolved(notation, limits)
	if err != nil {
		return nil, err
	}
	if fingerprint != plan.PlanFingerprint {
		return nil, newDiceError("UNSUPPORTED_NOTATION", "Roll plan does not match its source input", notation,
			map[string]any{"expectedFingerprint": plan.PlanFingerprint, "receivedFingerprint": fingerprint})
	}
	return plan, nil
}

func CompileRPGDice(input string, options ...CompileOptions) (*RollPlan, error) {
	return DefaultDiceEngine().Compile(input, options...)
}
func InspectRPGDiceNotation(input string, options ...CompileOptions) *DiceNotationInspection {
	return DefaultDiceEngine().Inspect(input, options...)
}
func VerifyRPGDiceNotation(input string, options ...CompileOptions) bool {
	return DefaultDiceEngine().Verify(input, options...)
}
func RollRPGDice(input any, options ...RollOptions) (*DiceRollResult, error) {
	return DefaultDiceEngine().Roll(input, options...)
}
func RollRPGDiceDetails(input any, options ...RollOptions) (*DiceRollDetails, error) {
	return DefaultDiceEngine().RollDetails(input, options...)
}
func RollRPGDiceSummary(input any, options ...RollOptions) (*DiceRollSummary, error) {
	return DefaultDiceEngine().RollSummary(input, options...)
}
